### ISSUE TITLE
Issue sending message to gamma agent server from external python script

### ISSUE THREAD

**Issue by gitting-around** [gama-platform/gama#3037]
Hi,

I am having some troubles in setting up a communication between Gama agents and an external app.
I am using the tcp protocol for this, and the model looks like [gaml-server.docx](https://github.com/gama-platform/gama/files/5288430/gaml-server.docx), whereas the python script run externally looks like 
[python-tcpserver.docx](https://github.com/gama-platform/gama/files/5288434/python-tcpserver.docx). Basically, I am creating a client agent that connects to the external python script, and a server agent that receives any reply from the external app and forwards it back to the client agent. The python script simply sends back the data as it got it, no change. 
While I can send a message from the Gama client to the external app, the other way around doesn't seem to be working. Sometimes, it looks like one message goes through, usually in the first steps of the simulation, and then nothing. 

Any ideas why I am running into this issue?
I am running this on a Mac Catalina, and I downloaded the version for Mac with jdk.

Thanks in advance!

**Comment by benoitgaudou** [comment]
This will not fix your issue, but could you check that the examples of TCP exchanges with an external app (using Processing) work correctly? 
(examples are  provided  in the model library)

Please also specify the GAMA  version you downloaded  (release  1.8, 1.8.1 ? continuous built ... )

**Comment by gitting-around** [comment]
Hi,

I am using release 1.8.1.

Regarding the examples, this is what I have in the plugin (not model) library 
![Screenshot 2020-09-28 at 16 05 23 (2)](https://user-images.githubusercontent.com/11443476/94444035-2ca2d580-01a6-11eb-999d-6b14e11a7b03.png)

It seems that the processing examples are for UDP. Do I need to write similar for tcp, or are the any template scripts somewhere. If the former it'll take sometime to get into it.

Thanks!

**Comment by gitting-around** [comment]
Hi again,

I tried the udp example and processing, since this was straightforward. It seems to be working, i.e., when I move the mouse around, the locations in gama change also. What can I do next? Is there some other test I can run with tcp, or some other example with the tcp and an external app?

Thanks!

**Comment by benoitgaudou** [comment]
I tried to  simplify the problem to test the GAMA connection to a TCP  Server. 
Thank to your code, I made 2 small program in Python (one Client, one Server) as Jupyter Notebook.

[GAMA - ClientServerTCP.ipynb.zip](https://github.com/gama-platform/gama/files/5321923/GAMA.-.ClientServerTCP.ipynb.zip)

Up to now, I was able to make the Server works, but not the Client....
Can someone with more skills have look and make it work?

**Comment by gitting-around** [comment]
Hi,

I tried with your code. In my case what works is a gama client sending messages to the external server. Whereas the other way around, from external client to gama server does not work.

Indeed when the external client sends a message, the following gets printed: `sending data back to the client - gama server
None`

Any two cents anyone?

Thanks!

**Comment by gitting-around** [comment]
I also tried using a template client and server from processing. I.e. processing client to gama server, where the client looks lie
`import processing.net.*; 

Client myClient; 
int dataIn; 
 
void setup() { 
  size(200, 200); 
  // Connect to the local machine at port 5204.
  // This example will not run if you haven't
  // previously started a server on this port.
  myClient = new Client(this, "127.0.0.1", 5204); 
} 
 
void draw() { 
  if (myClient.available() > 0) { 
    dataIn = myClient.read(); 
  } 
  background(dataIn); 
} 
`
as gotten from [here](https://www.processing.org/reference/libraries/net/Client.html)

as well as processing server with gama client, where the server looks like:
`import processing.net.*;

Server myServer;
int val = 0;

void setup() {
  size(200, 200);
  // Starts a myServer on port 5204
  myServer = new Server(this, 5204); 
}

void draw() {
  val = (val + 1) % 255;
  background(val);
  myServer.write(val);
}`
as gotten from [here](https://processing.org/reference/libraries/net/Server.html).

Neither worked. So any pointer as to what is the next step would be highly appreciated.

**Comment by RoiArthurB** [comment]
Hi all, good news : I've been able to make TCP communications in between GAMA and any Python scripts (and more) as server and client ! :)

## TL;DR
In the GAMA client, you should not add your client to any group (if you don't use that function on your server) so it will broadcast the message to everyone on the port and be catched by your server (see group doc [here](https://gama-platform.github.io/wiki/BuiltInSkills#join_group) and usage network documentation [here](https://gama-platform.github.io/wiki/UsingNetwork)).
So the GAMA Client creation looks like this now : 
```gaml
...
create Networking_Client number:3 {
	// replace the "localhost" address by the IP address of the other computer 
	do connect to: "localhost" protocol: "tcp_client" port: 3002 with_name: "Client";
	//REMOVE THE LINE => do join_group with_name:"client_group";
}
...
```

With this fix, the Jupyter Notebook server/client is working :)

## GAMA TCP's client pipeline

> To do all the tests, I've been using [`netcat`](https://en.wikipedia.org/wiki/Netcat) to do a very simple TCP Server/Client in verbose mode, and I tested it after with Python

GAMA TCP Client will work at 3 different time:
- First while the simulation's initialization (starting it and creating your TCP clients). It will broadcast on the port to try to connect himself to a server
  - GAMA will freeze and wait until it can connect to a TCP server during this phase
- Second when the simulation is running, GAMA client will broadcast on the server
  - Note that, for now, messages are sent in a XML format of the form : 
```xml
<ummisco.gama.network.common.CompositeGamaMessage>
	<unread>true</unread>
	<sender class="string">Server3</sender>
	<receivers class="string">client_group</receivers>
	<contents class="string">&lt;string&gt;I am Server Networking_server3 I give order to Clients&lt;/string&gt;</contents>
	<emissionTimeStamp>4878</emissionTimeStamp>
</ummisco.gama.network.common.CompositeGamaMessage>
```
- Finally, when you close the client, it will close the connection in between the client and the server (and close the server if you don't explicit configure it to stay alive)

Below, full server log of a communication with GAMA client
```
$ netcat -v -l -p 3001# Starting server listener
# Connection of the GAMA client
Connection from 127.0.0.1:59604
#  Identification on the network 
@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@Client

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@Client

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@ALL

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@ALL

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@Client

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@ALL

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@Client

@b@@r@@b@@r@/127.0.0.1:3002@b@@r@localhost/127.0.0.1:3002@b@@r@0@b@@r@ALL

# GAMA simulation running on 3 steps

@b@@r@Client@b@@r@server_group@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <sender class="string">Client</sender>@n@  <receivers class="string">server_group</receivers>@n@  <contents class="string">&lt;string&gt;Networking_Client0 at 0 sent to Server a message&lt;/string&gt;</contents>@n@  <emissionTimeStamp>0</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>

@b@@r@Client@b@@r@server_group@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <sender class="string">Client</sender>@n@  <receivers class="string">server_group</receivers>@n@  <contents class="string">&lt;string&gt;Networking_Client1 at 0 sent to Server a message&lt;/string&gt;</contents>@n@  <emissionTimeStamp>0</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>

@b@@r@Client@b@@r@server_group@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <sender class="string">Client</sender>@n@  <receivers class="string">server_group</receivers>@n@  <contents class="string">&lt;string&gt;Networking_Client2 at 0 sent to Server a message&lt;/string&gt;</contents>@n@  <emissionTimeStamp>0</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>

# Closing simulation in GAMA

read(net): Connection reset by peer
```

**Comment by gitting-around** [comment]
Hi,

My problem was not with the Gama client, as that was working in my python script, as well as what @benoitgaudou provided. What is not working for me is the Gama server, neither with my script, nor the other one. Please look at the first message for a full account.

**Comment by hqnghi88** [comment]
[TCP Server Example 2.zip](https://github.com/gama-platform/gama/files/5326846/TCP.Server.Example.2.zip)
Hi,
Here is the modif from your first message, and it seems to works. Please check if it fit your need.

**Comment by gitting-around** [comment]
Hi @hqnghi88,

Indeed this works now, I am able to send back the data to the gama server. I checked what you changed, and it seemed like the way the name of the server was declared messed it up.  

After trying this, I also tried to send to the gama server a message that I create from the python script. I followed the form that @RoiArthurB mentioned in the previous comment. (I am not sure whether this belongs here, or maybe I should create another issue).

From the python script I create the message as follows:
` data = "@b@@r@Client@b@@r@server_group@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <sender class=\"string\">Client</sender>@n@  <receivers class=\"string\">server_group</receivers>@n@  <contents class=\"string\">&lt;string&gt;bye bye bye&lt;/string&gt;</contents>@n@  <emissionTimeStamp>1</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>"
`
then I do `s.sendall(data.encode())`

I also tried with 
`data = "<ummisco.gama.network.common.CompositeGamaMessage><unread>true</unread><sender class=\"string\">Client</sender><receivers class=\"string\">server_group</receivers><contents class=\"string\">&lt;string&gt;bye bye bye&lt;/string&gt;</contents><emissionTimeStamp>1</emissionTimeStamp></ummisco.gama.network.common.CompositeGamaMessage>"
`

However, nothing is received. I guess I am building the message in a bad way. I also wonder whether the missionTimeStamp needs to be set in a certain way. 

Thank you for your effort!

**Comment by hqnghi88** [comment]
Hi,
Some trick to do with the encoding/decoding between gama and python, here is the string that works (dont know if github can show it correctly, or you can find in the file)
[external.zip](https://github.com/gama-platform/gama/files/5327725/external.zip)


                data = "@b@@r@Client@b@@r@server_group@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <sender class='string'>Client</sender>@n@  <receivers class='string'>server_group</receivers>@n@  <contents class='string'>&lt;string&gt;bye bye bye&lt;/string&gt;</contents>@n@  <emissionTimeStamp>8</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>\n"
and you need a newline in the string because gama will read the message line by line

**Comment by gitting-around** [comment]
This is amazing! Thank you, thank you, thank you!

I would sincerely advice you to put these snippets of code somewhere in the documentation to help people get started :).

I will now go ahead and close the issue, thanks again, and have a great day!
